proof/refine/ARM/KHeap_R.thy

```diff
@@ -4177,4 +4177,10 @@ lemma sym_heap_remove_only:
   apply (subst (asm) sym_heap_symmetric[simplified sym_heap_def], simp)
   done
 
+lemma pred_tcb_at'_pred_map:
+  "pred_tcb_at' p P t s = (tcb_at' t s \<and> pred_map (\<lambda>x. P (p (tcb_to_itcb' x))) (tcbs_of' s) t)"
```

Comment:
Gerenally speaking, I think we can do without `pred_map` here. Something like`tcb_at' t s \<and> P ((tcbs_of’ s |> p o tcb_to_itcb) t)` for RHS should work.

Reply:
Will change it.

Reply:
Hmm. I just saw what you actually replaced this with. Do you prefer `the` to `opt_map` for a particular reason? I would think that using `opt_map` goes better with the general projections direction. Not sure if we really want to use `the` here.

Reply:
Eventually, it would be nice to get rid of `itcb` altogether and replace it with projections.

Reply:
The opt-map version wasn't quite working for me in the case where p does not return an option type. This is relevant for the case where p is `tcbState` or similar. I would be very happy with an opt_map version that covers that situation, if possible. Do you think that is possible?

Reply:
Is there a way to get the opt-map version working in that case?

Reply:
Ah, yes, of course, that won’t work. Sorry, a bit of overboard. Still, we do have a bit of duplications now, like `itcbSchedContext` and ‘tcbSCs_of’. Maybe we could remove `obj_ref` fields of tcb from itcb one day.